File: 04 Polinomial Regression/polinomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Thu Dec 19 20:06:20 2019

@author: Daniel
"""
# Regresión Polinómica

# Como importar librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Importar dataset
dataset = pd.read_csv('Position_Salaries.csv')
#Iloc Sirve para localizar elemento de un dataset
X = dataset.iloc[:, 1:2].values 
y = dataset.iloc[:, 2].values

# Dividir el data set en conjunto de entrenamiento y conjunto de testing
"""from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, 
random_state = 0)"""

# Escalado de variables
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)"""

# Ajustar la regresión lineal con el dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y) 

# Ajustar la regresión polinómica con el dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

# Visualización de los resultados del modelo lineal
plt.scatter(X, y, color = "red")
plt.plot(X, lin_reg.predict(X), color = "blue")
plt.title("Modelo de regresión lineal")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo en ($)")
plt.show()

# Visualización de los resultados del modelo polinómico
X_grid = np.arange(min(X), max(X),0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color = "red")
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), 
         color = "blue")
plt.title("Modelo de regresión polinómica")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo en ($)")
plt.show()

# Predicción de nuestros modelos
lin_reg.predict([[6.5]])
lin_reg_2.predict(poly_reg.fit_transform([[6.5]]))

File: 04 Polinomial Regression/polinomial_regression.R
# Regresion Polinomica

# Importar el dataset
dataset = read.csv('04 Polinomial Regression/Position_Salaries.csv')
dataset = dataset[, 2:3]

# Ajustar modelo de regresión lineal con el conjunto de datos
lin_reg = lm(formula = Salary~.,
             data = dataset)

# Ajustar modelo de regresión polinómica con el conjunto de datos
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4

poly_reg = lm(formula = Salary~.,
              data = dataset)

# Visualización del modelo lineal
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = "red") + 
  geom_line(aes(x =dataset$Level, 
                y =predict(lin_reg, newdata = dataset)),
            color = "blue") +
  ggtitle("Predicción lineal del suelo en función del nivel del empleado")+
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")

# Visualización del modelo polinómico
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = "red") + 
  geom_line(aes(x =dataset$Level, 
                y =predict(poly_reg, newdata = dataset)),
            color = "blue") +
  ggtitle("Predicción lineal del suelo en función del nivel del empleado")+
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")

# Prediccion de nuevos resultados con Regresion Lineal 
y_pred = predict(lin_reg, newdata = data.frame(Level = 6.5))

# Prediccion de nuevos resultados con Regresión Polinómica
y_pred = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                Level2 = 6.5^2,
                                                Level3 = 6.5^3,
                                                Level4 = 6.5^4))



